File: frontend/src/types/course.ts
export interface Lesson {
  _id: string;
  id?: string; // Pour la rétrocompatibilité
  title: string;
  description?: string;
  content: string;
  duration: number;
  type: 'video' | 'text' | 'quiz';
  videoUrl?: string; // URL de la vidéo (YouTube, Vimeo, etc.)
  video_url?: string; // Format backend pour l'URL de la vidéo
  attachments?: Array<{
    name: string;
    url: string;
    type?: string;
    size?: number;
  }>; // Fichiers joints à la leçon (PDF, etc.)
  order: number;
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

export interface Module {
  _id: string;
  id?: string; // Pour la rétrocompatibilité
  title: string;
  description?: string;
  order: number;
  lessons: Lesson[];
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

export type CourseLevel = "beginner" | "intermediate" | "advanced";

export interface Course {
  _id: string;
  id?: string; // Pour la rétrocompatibilité
  title: string;
  description: string;
  category_id: string;
  category?: string; // Nom de la catégorie
  price: number;
  is_active: boolean;
  featured: boolean;
  level: CourseLevel;
  duration?: number | string; // Peut être un nombre ou une chaîne comme "2h30min"
  language: string;
  prerequisites?: string[];
  objectives?: string[];
  slug: string;
  modules: Module[];
  enrolled_students: number;
  enrollmentCount?: number; // Alias pour enrolled_students
  instructor?: string; // Nom de l'instructeur
  coverImage?: string; // URL de l'image de couverture
  rating: number;
  total_ratings: number;
  created_at: string;
  updated_at: string;
  createdAt?: string; // Alias pour created_at
} 

File: backend/app/models/course.py
from datetime import datetime
from typing import Optional, List, Any
from pydantic import BaseModel, Field, GetJsonSchemaHandler
from pydantic.json_schema import JsonSchemaValue
from bson import ObjectId
from .course_category import PyObjectId, generate_slug

class Lesson(BaseModel):
    title: str
    description: Optional[str] = None
    content: str
    duration: int  # en minutes
    type: str = "video"  # video, text, quiz, etc.
    order: int
    video_url: Optional[str] = None  # URL de la vidéo pour les leçons de type video
    is_active: bool = True
    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime = Field(default_factory=datetime.utcnow)

    class Config:
        json_encoders = {ObjectId: str}

class Module(BaseModel):
    title: str
    description: Optional[str] = None
    order: int
    lessons: List[Lesson] = []
    is_active: bool = True
    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime = Field(default_factory=datetime.utcnow)

    class Config:
        json_encoders = {ObjectId: str}

class CourseBase(BaseModel):
    title: str
    description: str
    category_id: PyObjectId
    price: float = 0.0
    is_active: bool = True
    featured: bool = False
    level: str = "beginner"  # beginner, intermediate, advanced
    duration: Optional[int] = None  # durée totale en minutes
    language: str = "fr"
    prerequisites: Optional[List[str]] = None
    objectives: Optional[List[str]] = None

class CourseCreate(CourseBase):
    slug: Optional[str] = None
    modules: List[Module] = []

    def __init__(self, **data):
        super().__init__(**data)
        if not self.slug:
            self.slug = generate_slug(self.title)
        if not self.duration:
            # Calculer la durée totale à partir des modules et leçons
            total_duration = sum(
                sum(lesson.duration for lesson in module.lessons)
                for module in self.modules
            )
            self.duration = total_duration

class CourseUpdate(BaseModel):
    title: Optional[str] = None
    description: Optional[str] = None
    category_id: Optional[PyObjectId] = None
    price: Optional[float] = None
    is_active: Optional[bool] = None
    featured: Optional[bool] = None
    level: Optional[str] = None
    duration: Optional[int] = None
    language: Optional[str] = None
    prerequisites: Optional[List[str]] = None
    objectives: Optional[List[str]] = None
    slug: Optional[str] = None
    modules: Optional[List[Module]] = None

class CourseInDB(CourseBase):
    id: PyObjectId = Field(default_factory=PyObjectId, alias="_id")
    slug: str
    modules: List[Module] = []
    enrolled_students: int = 0
    rating: float = 0.0
    total_ratings: int = 0
    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime = Field(default_factory=datetime.utcnow)

    class Config:
        allow_population_by_field_name = True
        arbitrary_types_allowed = True
        json_encoders = {ObjectId: str}
        schema_extra = {
            "example": {
                "title": "Développement Web Full Stack",
                "description": "Apprenez à créer des applications web modernes",
                "category_id": "507f1f77bcf86cd799439011",
                "price": 99.99,
                "is_active": True,
                "featured": False,
                "level": "beginner",
                "duration": 1200,
                "language": "fr",
                "prerequisites": ["Bases de l'informatique"],
                "objectives": ["Créer une application web complète"],
                "slug": "developpement-web-full-stack",
                "modules": [],
                "enrolled_students": 0,
                "rating": 0.0,
                "total_ratings": 0,
                "created_at": "2024-01-15T10:00:00",
                "updated_at": "2024-01-15T10:00:00"
            }
        } 